modules/assim.sequential/R/downscale_function.R

```diff
@@ -140,84 +140,141 @@ SDA_downscale <- function(preprocessed, date, carbon_pool, covariates, model_typ
       predictions[[i]] <- stats::predict(models[[i]], test_data)
     }
   } else if (model_type == "cnn") {
+    # Define k_folds and num_bags
+    k_folds <- 5
+    num_bags <- 5
+
     # Reshape input data for CNN
     x_train <- keras3::array_reshape(x_train, c(nrow(x_train), 1, ncol(x_train)))
     x_test <- keras3::array_reshape(x_test, c(nrow(x_test), 1, ncol(x_test)))
-    
+
     for (i in seq_along(carbon_data)) {
-      # Define the CNN model architecture
-      # Used dual batch normalization and dropout as the first set of batch normalization and dropout operates on the lower-level features extracted by the convolutional layer, the second set works on the higher-level features learned by the dense layer.
-      model <- keras3::keras_model_sequential() |>
-        # 1D Convolutional layer: Extracts local features from input data
-        keras3::layer_conv_1d(filters = 64, kernel_size = 1, activation = 'relu', input_shape = c(1, length(covariate_names))) |>
-        # Batch normalization: Normalizes layer inputs, stabilizes learning, reduces internal covariate shift
-        keras3::layer_batch_normalization() |>
-        # Dropout: Randomly sets some of inputs to 0, reducing overfitting and improving generalization
-        keras3::layer_dropout(rate = 0.3) |>
-        # Flatten: Converts 3D output to 1D for dense layer input
-        keras3::layer_flatten() |>
-        # Dense layer: Learns complex combinations of features
-        keras3::layer_dense(units = 64, activation = 'relu') |>
-        # Second batch normalization: Further stabilizes learning in deeper layers
-        keras3::layer_batch_normalization() |>
-        # Second dropout: Additional regularization to prevent overfitting in final layers
-        keras3::layer_dropout(rate = 0.3) |>
-        # Output layer: Single neuron for regression prediction
-        keras3::layer_dense(units = 1)
+      all_models <- list()
       
-      # Learning rate scheduler
-      lr_schedule <- keras3::learning_rate_schedule_exponential_decay(
-        initial_learning_rate = 0.001,
-        decay_steps = 1000,
-        decay_rate = 0.9
-      )
+      # Create k-fold indices
+      fold_indices <- caret::createFolds(y = 1:nrow(x_train), k = k_folds, list = TRUE, returnTrain = FALSE)
```

Comment:
This introduces a new dependency for what is actually appears to be a fairly straightforward task. Are there analogous functions you could use from packages that are already in the dependencies? Could you do this manually with base functions like `sample` and/or a bit of tidyverse?  If the answers to this are no, you need to add the new package to the dependencies and update the PR.

Reply:
FWIW the 'tidyverse' equivalent would be `rsample::vfold_cv`, but still might not be worth adding this dependency since implementation "should be" 🤞🏻 simple, like:

```
# <incomplete function documentation>
# @param i vector of indices created by seq_along(x)
# @param k number of groups
# based on https://stackoverflow.com/a/57085061/199217
create_folds <- function(i, k) {
  indices <- sample(i)
  folds <- split(indices, cut(indices, breaks = k, labels = FALSE))
  return(folds)
}
```

Ok it wasn't as simple as I hoped, but that should work!

Reply:
Addressed in the commits [4ae1974](https://github.com/PecanProject/pecan/pull/3357/commits/4ae1974d7aa78fa27e188d222e85ab0eba14b773) , [5d86312](https://github.com/PecanProject/pecan/pull/3357/commits/5d86312d92fe03adeb3a77665b96db4ec2dcda45) and [70bfdaf](https://github.com/PecanProject/pecan/pull/3357/commits/70bfdaf5ac40a69309d78b063c28d41576d5603d)